### PR TITLE
APIGOV-17556 - add sampling package for traceability

### PR DESCRIPTION
**Pull request by jcollins-axway**
Would like any feedback now as I work to update the traceability agents to use the new CreateTransactionEvents method in event generator

### CODE REVIEW

File: pkg/traceability/sampling/sample.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+func (s *sample) ShouldSampleTransaction(details TransactionDetails) bool {
+	// Only sampling on percentage, not currently looking at the details
+	samplePercent := s.config.Percentage
+	shouldSample := false
+	if s.currentCount < samplePercent {
+		shouldSample = true
+	}
+	s.currentCount++
+
+	// resent te count once we hit 100 messages
```
Comment (dfeldick) [inline pkg/traceability/sampling/sample.go:21]: reset the count

File: pkg/traceability/sampling/globalsampling.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+//DefaultConfig - returns a default sampling config where all transactions are sent
+func DefaultConfig() Sampling {
+	return Sampling{
+		Percentage: 100,
+	}
+}
+
+// SetupSampling - set up redactionRegex based on the redactionConfig
```
Comment (dfeldick) [inline pkg/traceability/sampling/globalsampling.go:24]: this isn't the right comment

File: pkg/traceability/sampling/sample.go (hunk 1 above)
Comment (jcollins-axway) [reply pkg/traceability/sampling/sample.go]: fixed

File: pkg/traceability/sampling/sample.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+// sample - private struct that is used to keep track of the samples being taken
+type sample struct {
+	config       Sampling
+	currentCount int
+}
+
+// ShouldSampleTransaction - receives the transaction details and returns true to sample it false to not
+func (s *sample) ShouldSampleTransaction(details TransactionDetails) bool {
+	// Only sampling on percentage, not currently looking at the details
+	samplePercent := s.config.Percentage
```
Comment (dfeldick) [inline pkg/traceability/sampling/sample.go:14]: really don't need the var samplePercent

Comment (jcollins-axway) [reply pkg/traceability/sampling/sample.go:14]: removed, thanks
\

File: pkg/traceability/sampling/globalsampling.go (hunk 2 above)
Comment (jcollins-axway) [reply pkg/traceability/sampling/globalsampling.go:24]: fixed

**Review by dfeldick — APPROVED** [review]

File: pkg/transaction/eventgenerator.go [hunk 4]
```diff
@@ -66,23 +84,58 @@ func (e *Generator) CreateEvent(logEvent LogEvent, eventTime time.Time, metaData
...
+	return beat.Event{
 		Timestamp: eventTime,
 		Meta:      metaData,
 		Private:   privateData,
 		Fields:    eventData,
+	}, nil
+}
+
+// CreateTransactionEvents - Creates new events to be sent to Condor
+func (e *Generator) CreateTransactionEvents(summaryEvent LogEvent, detailEvents []LogEvent, eventTime time.Time, metaData common.MapStr, eventFields common.MapStr, privateData interface{}) ([]beat.Event, error) {
```
Comment (vivekschauhan) [inline pkg/transaction/eventgenerator.go:99]: The method name sounds a bit obscure as it generates beats event wrapping both transaction summary and transaction events and not only transaction events. Can we  generalize the name to may be CreateEvents /CreateSampledEvents or something similar?

**Review by vivekschauhan — APPROVED** [review]

File: pkg/transaction/eventgenerator.go (hunk 4 above)
Comment (jcollins-axway) [reply pkg/transaction/eventgenerator.go]: CreateEvents, done

**Review by dfeldick — APPROVED** [review]

**Review by dfeldick — APPROVED** [review]

**Review by dfeldick — APPROVED** [review]

**Review by dfeldick — APPROVED** [review]

**Pull request merged**